File: .claude/skills/mt-add-url/scripts/find-newsletter-number.js
#!/usr/bin/env node
// Find the most recent newsletter number and return the next one
// Usage: node find-newsletter-number.js
// Outputs: The next newsletter number

const fs = require("fs");
const path = require("path");

const PROJECT_ROOT = path.resolve(__dirname, "../../../..");
const CONTENT_DIR = path.join(PROJECT_ROOT, "content", "post");

// Extract newsletter number from file content
function extractNewsletterNumber(filePath) {
  try {
    const content = fs.readFileSync(filePath, "utf-8");
    const match = content.match(/Newsletter\s*#(\d+)/);
    return match ? parseInt(match[1], 10) : 0;
  } catch {
    return 0;
  }
}

// Scan all year/month/day directories for newsletter posts
function findMostRecentNewsletter() {
  let maxNumber = 0;

  // List year directories, sorted descending
  let years;
  try {
    years = fs.readdirSync(CONTENT_DIR)
      .filter((d) => /^\d{4}$/.test(d))
      .sort((a, b) => b.localeCompare(a));
  } catch {
    return 0;
  }

  for (const year of years) {
    const yearDir = path.join(CONTENT_DIR, year);
    let months;
    try {
      months = fs.readdirSync(yearDir)
        .filter((d) => /^\d{2}$/.test(d))
        .sort((a, b) => b.localeCompare(a));
    } catch {
      continue;
    }

    for (const month of months) {
      const monthDir = path.join(yearDir, month);
      let days;
      try {
        days = fs.readdirSync(monthDir)
          .filter((d) => /^\d{2}$/.test(d))
          .sort((a, b) => b.localeCompare(a));
      } catch {
        continue;
      }

      for (const day of days) {
        const filePath = path.join(monthDir, day, "index.md");
        const num = extractNewsletterNumber(filePath);
        if (num > maxNumber) maxNumber = num;
      }
    }

    // Early exit: if we found a newsletter in this year, no need to go further back
    if (maxNumber > 0) break;
  }

  return maxNumber;
}

const mostRecent = findMostRecentNewsletter();
console.log(mostRecent + 1);

File: scripts/newsletter/find_newsletter_number.go
// Find the most recent newsletter number and return the next one.
// Usage: go run ./scripts/newsletter find-newsletter-number
// Outputs: the next newsletter number.
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
)

var (
	newsletterNumRe = regexp.MustCompile(`Newsletter\s*#(\d+)`)
	yearDirRe       = regexp.MustCompile(`^\d{4}$`)
	twoDigitDirRe   = regexp.MustCompile(`^\d{2}$`)
)

// listDirsDesc returns dir's subdirectory names matching re, sorted descending.
func listDirsDesc(dir string, re *regexp.Regexp) []string {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil
	}
	var names []string
	for _, e := range entries {
		if re.MatchString(e.Name()) {
			names = append(names, e.Name())
		}
	}
	sort.Sort(sort.Reverse(sort.StringSlice(names)))
	return names
}

func extractNewsletterNumber(path string) int {
	content, err := os.ReadFile(path)
	if err != nil {
		return 0
	}
	m := newsletterNumRe.FindSubmatch(content)
	if m == nil {
		return 0
	}
	n, err := strconv.Atoi(string(m[1]))
	if err != nil {
		return 0
	}
	return n
}

// findMostRecentNewsletter scans year/month/day directories newest-first for
// the highest newsletter number.
func findMostRecentNewsletter() int {
	maxNumber := 0
	for _, year := range listDirsDesc(contentDir(), yearDirRe) {
		yearDir := filepath.Join(contentDir(), year)
		for _, month := range listDirsDesc(yearDir, twoDigitDirRe) {
			monthDir := filepath.Join(yearDir, month)
			for _, day := range listDirsDesc(monthDir, twoDigitDirRe) {
				if n := extractNewsletterNumber(filepath.Join(monthDir, day, "index.md")); n > maxNumber {
					maxNumber = n
				}
			}
		}
		// Early exit: a newsletter found in this year — no need to go further back.
		if maxNumber > 0 {
			break
		}
	}
	return maxNumber
}

func runFindNewsletterNumber(_ []string) {
	fmt.Println(findMostRecentNewsletter() + 1)
}
